impl `Sync` for `SerialPort` trait

unlimitedsola opened the pull request:
Closes #232 

There are many alternatives that don't involve `unsafe impl`, but I believe this is the simplest one that doesn't involve an MSRV bump, nor should it introduce breakage, at least I hope...

One alternative is to use [`OwnedHandle`](https://doc.rust-lang.org/std/os/windows/io/struct.OwnedHandle.html) provided by stdlib. However, doing so requires an MSRV bump to at least 1.63.0, which I've heard would require a major version bump in this library so I refrained from this approach.

unlimitedsola committed:
impl `Sync` for `SerialPort` trait

sirhcel commented on src/windows/com.rs:
```diff
@@ -32,7 +32,17 @@ pub struct COMPort {
     port_name: Option<String>,
 }
 
+// The Windows [`HANDLE`] type is considered safe according to the standard library.
+// See explanation on [`OwnedHandle`] in stdlib for more information.
```
Looking at https://doc.rust-lang.org/std/os/windows/raw/type.HANDLE.html or https://doc.rust-lang.org/std/os/windows/io/struct.OwnedHandle.html, I don't see the documentation you are mentioning at al first glance. Where is the statement made you are referring to? In [this comment in `handle.rs`](https://doc.rust-lang.org/src/std/os/windows/io/handle.rs.html#111)

unlimitedsola replied:
Yes, I'm referring to the comment in [`handle.rs`](https://doc.rust-lang.org/src/std/os/windows/io/handle.rs.html#111-117) that you linked.
I'm not sure how to link it to an `impl` block so I figured the definition of `OwnedHandle` might be the closest to that comment (50+ lines away)

The link to the std source https://doc.rust-lang.org/src/std/os/windows/io/handle.rs.html#111-117 doesn't seem to be a permanent one so I was worried putting the doc link in might become invalid in the future.

I can probably replace it with a GitHub permalink like this: https://github.com/rust-lang/rust/blob/f4f0fafd0c7849e162eddbc69fa5fe82dbec28c7/library/std/src/os/windows/io/handle.rs#L111-L115
but that would be a very long line unfortunately.

sirhcel replied:
Alight. A permalink would be great then because a long line looks way more appealing to me than the risk of getting it broken at some point in time.

unlimitedsola replied:
Thanks for your feedback! I've now updated the patch with a permalink to the explanation.

sirhcel reviewed (commented):
Thank you very much for this PR! Please see my comment below. I'm wondering what's you use case for sharing `&SerialPort` between threads? At a firs glance there are only a few meaningful things to do with an immutable reference.

unlimitedsola commented:
> I'm wondering what's you use case for sharing `&SerialPort` between threads? At a firs glance there are only a few meaningful things to do with an immutable reference.

I was trying to display some port information and settings that should only require an immutable reference in another thread. Without `impl Sync`, however, it will force me to pass _ownership_ to that thread and then get it back.

unlimitedsola committed:
Provide a permalink to the explanation in stdlib

okhsunrog commented:
Is there anything that blocks this from getting merged?

okhsunrog commented:
@sirhcel I used this crate via tokio-serial crate, and it worked fine on Linux. Once I tried to compile my app for Windows, I saw the error that SeriaPort is not Sync on windows. So, for now I have this in my code:
```rust
    // SerialStream from tokio-serial crate is not Sync on Windows, so we need to spawn a separate thread for mmw_lib
    #[cfg(windows)]
    {
        // Create a new single-threaded runtime for Windows
        let runtime = tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()
            .unwrap();

        // Spawn the radar connection handler on the new runtime
        std::thread::spawn(move || {
            runtime.block_on(async {
                radar_connection_handler(rx_radar_control).await;
            });
        });
    }

    #[cfg(not(windows))]
    {
        // Spawn a new tokio task to manage radar connection (original behavior for non-Windows platforms)
        tokio::spawn(async move {
            radar_connection_handler(rx_radar_control).await;
        });
    }
```

okhsunrog commented:
@sirhcel so the answer is: it can be only used in a single-threaded tokio runtime on Windows, sadly

unlimitedsola commented:
> I already added semver checks to CI with #240 but I'm apparently not entitled to rebase this PR on the current upstream master. @unlimitedsola, could you please rebase this PR on upstream/master or allow edits by maintainers?

Rebased, please take a look.

sirhcel commented:
Thank you for rebasing @unlimitedsola! The changes proposed here are semver-breaking - at least in theory.

Therefor I would prefer checking the our visible dependents from crates.io and GitHub whether they would be affected by this change first. I did this already in the past and have to dust off the makeshift setup used there. But this will take some time.

okhsunrog commented:
@sirhcel is there anything stopping it from being merged right now?

sirhcel commented:
> @sirhcel is there anything stopping it from being merged right now?

Yes, implementing `Sync` is actually a semver-breaking change. Cutting a release 5.0 just for this change seems a bit to much of semver compliance to me because I bet that this change does not affect depentents in practice. However, I want to give it a spin at least with our most-prominent dependents but I did not find the time for it the last weeks. ☹️

I don't know of tooling for this purpose (like a crater for libs) and used a [modified get-all-crates](https://github.com/sirhcel/get-all-crates/tree/dependents-filter) the last time. But I killed all the scripting I created back then when deleting the server. I need to find some time to set it up again. Any hints to prior art are welcome @okhsunrog.

sirhcel commented:
I picked the changes into [playground-5.0](https://github.com/serialport/serialport-rs/tree/playground-5.0).

unlimitedsola commented:
I think we can close this now as we have decided that we will adopt the approach in #327. Thanks for following up on this!
